gnmidiff/json.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2023 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gnmidiff
+
+import (
+	"encoding/json"
+	"fmt"
+	"sort"
+
+	"github.com/openconfig/ygot/ygot"
+)
+
+// flattenOCJSON outputs all leaf path-value pairs in the root per RFC7951.
+//
+// It assumes the input JSON is with respect to OpenConfig-compliant YANG.
+// See the following for checking compliance:
+// * https://github.com/openconfig/oc-pyang
+// * https://github.com/openconfig/public/blob/master/doc/openconfig_style_guide.md
+//
+// Output path format is per gNMI's path conventions.
+//
+// e.g.
+// {
+//   "openconfig-network-instance:config": {
+//     "description": "VRF RED",
+//     "enabled": true,
+//     "enabled-address-families": [
+//       "openconfig-types:IPV4",
+//       "openconfig-types:IPV6"
+//     ],
+//     "name": "RED",
+//     "type": "openconfig-network-instance-types:L3VRF"
+//     },
+//   "openconfig-network-instance:name": "RED"
+// }
+//
+// returns
+// {
+//   "openconfig-network-instance:config/description": "VRF RED",
+//   "openconfig-network-instance:config/enabled": true,
+//   "openconfig-network-instance:config/enabled-address-families": ["openconfig-types:IPV4", "openconfig-types:IPV6"],
+//   "openconfig-network-instance:config/name": "RED",
+//   "openconfig-network-instance:config/type": "openconfig-network-instance-types:L3VRF",
+//   "openconfig-network-instance:name": "RED",
+// }
+func flattenOCJSON(json7951 []byte) (map[string]interface{}, error) {
```

Comment:
Curious: given that you'd like `<path, value>` tuples, why not unmarshal this and then serialise it with `TogNMINotifications`? This way you don't end up with fragility around the particular YANG schema that you're dealing with AFAICS?

Sure I'm missing something, but I couldn't quite tell why this was the preferred implementation.

Reply:
The key extraction seems fragile -- but I agree that if you want to work with no knowledge of the schema, then there is little choice.

Reply:
Makes sense, I'll make using schema as the default behaviour, and add a flag to invoke this implementation.